docs: Add a user guide on ScrapeConfig (#5659)
prometheus-operator/prometheus-operator: Documentation/user-guides/scrapeconfig.md

labels to match `ScrapeConfigs`:

```yaml
spec:
  scrapeConfigSelector:
    prometheus: system-monitoring-prometheus
```

With this example, all `ScrapeConfig` having the `prometheus` label set to `system-monitoring-prometheus` will be used
to generate scrape configurations.

# Use ScrapeConfig to scrape an external target

`ScrapeConfig` currently supports a limited set of service discoveries:
* `static_config`
* `file_sd`
* `http_sd`

Also, the following examples are basic and the CRD is constantly evolving, adding new features. Check the API documentation to see all the available fields.

If you have an interest in another service discovery mechanism or you see something missing in the implementation, please
[open an issue](https://github.com/prometheus-operator/prometheus-operator/issues).

## `static_config`

For example, to scrape a target located at `http://10.1.2.3:9100`, use the following:

```yaml
apiVersion: coreos.monitoring.com/v1alpha1
kind: ScrapeConfig
metadata:
  name: my-scrape-config
  namespace: my-namespace
spec:
  staticConfigs:
    - targets: 

>       - 10.1.2.3:9100

Reviewer: add label too?

```

## `file_sd`

To use `file_sd`, a file has to be mounted in the Prometheus or PrometheusAgent pods. The following configmap is a service discovery file:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: scrape-file-sd-targets
  namespace: monitoring
data:
  targets.yaml: |
    - labels:
        job: node-demo
      targets:
      - node.demo.do.prometheus.io
    - labels:
        job: prometheus
      targets:

>       - aremoteprometheus.io:9090

Reviewer: we could use prometheus.demo.do.prometheus.io

```

This `ConfigMap` will then need to be mounted in the `Prometheus` spec:

```yaml
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: your-prometheus
  namespace: monitoring
spec:

>   [...]

Reviewer: probably add configmap section here?

```yaml
spec:
  configMaps:
   - scrape-file-sd-targets
```

Reviewer: Also we would need to add `scrapeConfigSelector: {}`. It doesn't work without adding that

---
weight: 153
toc: true
title: ScrapeConfig CRD
menu:
    docs:
        parent: user-guides
lead: ""
images: []
draft: false
description: Guide to use ScrapeConfig to scrape targets external to the Kubernetes cluster
---

Starting with prometheus-operator v0.65.x, one can use the `ScrapeConfig` CRD to scrape targets external to the
Kubernetes cluster or create scrape configurations that are not possible with the higher level
`ServiceMonitor`/`Probe`/`PodMonitor` resources.

# Prerequisites
* `prometheus-operator` `>v0.65.1`
* `ScrapeConfig` installed in the cluster. Make sure to restart the operator after having pushed the CRD.

> ## Configure Prometheus or PrometheusAgent to select ScrapeConfigs

Reviewer 2: (nit) 
```suggestion
# Configure Prometheus or PrometheusAgent to select ScrapeConfigs
```

Author: This was for me part of the prerequisites but I'm not too fussed either way

Both the Prometheus and PrometheusAgent CRD have a `scrapeConfigSelector` field. This field needs to be set to a list of
labels to match `ScrapeConfigs`:

```yaml
spec:
  scrapeConfigSelector:
    prometheus: system-monitoring-prometheus
```

With this example, all `ScrapeConfig` having the `prometheus` label set to `system-monitoring-prometheus` will be used
to generate scrape configurations.

# Use ScrapeConfig to scrape an external target

`ScrapeConfig` currently supports a limited set of service discoveries:
* `static_config`
* `file_sd`
* `http_sd`
* `kubernetes_sd`
* `consul_sd`

The following examples are basic and don't cover all the supported service discovery mechanisms. The CRD is constantly evolving, adding new features. Check the API documentation to see all the available fields.

If you have an interest in another service discovery mechanism or you see something missing in the implementation, please
[open an issue](https://github.com/prometheus-operator/prometheus-operator/issues).

## `static_config`

> For example, to scrape a target located at `http://10.1.2.3:9100`, use the following:

Reviewer 2: ```suggestion
For example, to scrape the target located at `http://prometheus.demo.do.prometheus.io:9090`, use the following:
```

```yaml
apiVersion: coreos.monitoring.com/v1alpha1
kind: ScrapeConfig
metadata:

>   name: my-scrape-config

Reviewer 2: (nit)
```suggestion
  name: static-config
```

  namespace: my-namespace
spec:
  staticConfigs:
    - labels:
        job: prometheus
      targets:
        - prometheus.demo.do.prometheus.io:9090
```

## `file_sd`

To use `file_sd`, a file has to be mounted in the Prometheus or PrometheusAgent pods. The following configmap is a service discovery file:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: scrape-file-sd-targets
  namespace: monitoring
data:
  targets.yaml: |
    - labels:
        job: node-demo
      targets:
      - node.demo.do.prometheus.io:9100
    - labels:
        job: prometheus
      targets:
      - prometheus.demo.do.prometheus.io:9090
```

This `ConfigMap` will then need to be mounted in the `Prometheus` spec:

```yaml
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: your-prometheus
  namespace: my-namespace
spec:
  scrapeConfigSelector:
    prometheus: system-monitoring-prometheus
  configMaps:
    - scrape-file-sd-targets
```

You can then use ScrapeConfig to reference that file and scrape the associated targets:

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: ScrapeConfig
metadata:

>   name: scrape-config-example

Reviewer 2: ```suggestion
  name: file-sd
```

lead: ""
images: []
draft: true
description: Guide to use ScrapeConfig to scrape targets external to the Kubernetes cluster
---

Starting with prometheus-operator v0.65.x, one can use the `ScrapeConfig` CRD to scrape targets external to the
Kubernetes cluster or create scrape configurations that are not possible with the higher level
`ServiceMonitor`/`Probe`/`PodMonitor` resources.

# Prerequisites
* `prometheus-operator` `>v0.65.1`
* `ScrapeConfig` installed in the cluster. Make sure to restart the operator after having pushed the CRD.

## Configure Prometheus or PrometheusAgent to select ScrapeConfigs

Both the Prometheus and PrometheusAgent CRD have a `scrapeConfigSelector` field. This field needs to be set to a list of
labels to match `ScrapeConfigs`:

```yaml
spec:
  scrapeConfigSelector:
    prometheus: system-monitoring-prometheus
```

With this example, all `ScrapeConfig` having the `prometheus` label set to `system-monitoring-prometheus` will be used
to generate scrape configurations.

# Use ScrapeConfig to scrape an external target

`ScrapeConfig` currently supports a limited set of service discoveries:
* `static_config`
* `file_sd`
* `http_sd`

> * `kubernetes_sd`

Reviewer 3: Do we plan to document all fields added to ScrapeConfig?

I'm a bit afraid that we could easily miss documentation as time goes on. Providing examples for all service discovery options might make this page quite big 🤔

Author: I was actually thinking about that and wanted to discuss having some use case based documentation rather than this (e.g. how to scrape exporters on the kubernetes nodes or how to scrape an external target etc). That way we don't have to document everything in one page but can focus on use cases that drove the implementation.

What do you think?

Reviewer 3: Since we don't have any documentation and people are already asking for guidance on ScrapeConfig, we could merge while the CRD is still small. We could rethink and rewrite it later.

I was raising this question mostly because I was concerned that we would have to have this discussion at some point 🤔

Reviewer 2: I'd be +1 on having the documentation around specific use cases. And I feel that the current examples fulfill the requirement without going into too many details.

Author: Alright, I'll keep it at that for now and we can figure out use case based docs as we go 👍

Also, the following examples are basic and the CRD is constantly evolving, adding new features. Check the API documentation to see all the available fields.

If you have an interest in another service discovery mechanism or you see something missing in the implementation, please
[open an issue](https://github.com/prometheus-operator/prometheus-operator/issues).

## `static_config`

For example, to scrape a target located at `http://10.1.2.3:9100`, use the following:

```yaml
apiVersion: coreos.monitoring.com/v1alpha1
kind: ScrapeConfig
metadata:
  name: my-scrape-config
  namespace: my-namespace
spec:
  staticConfigs:
    - labels:
        job: mytarget
      targets: 
      - 10.1.2.3:9100
    - labels:
        job: prometheus
      targets:

>         - prometheus.demo.do.prometheus.io:9090

Reviewer 2: I like that we use the Prometheus demo instance as it's working out of the box. Do we need the other `10.1.2.3:9100` target?

Author: We don't, removed it 👍

```

## `file_sd`

To use `file_sd`, a file has to be mounted in the Prometheus or PrometheusAgent pods. The following configmap is a service discovery file:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: scrape-file-sd-targets
  namespace: monitoring
data:
  targets.yaml: |
    - labels:
        job: node-demo
      targets:
      - node.demo.do.prometheus.io:9100
    - labels:
        job: prometheus
      targets:
      - prometheus.demo.do.prometheus.io:9090
```

This `ConfigMap` will then need to be mounted in the `Prometheus` spec:

```yaml
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: your-prometheus
  namespace: monitoring
spec:
  scrapeConfigSelector:
    prometheus: system-monitoring-prometheus
  configMaps:
    - scrape-file-sd-targets
```

You can then use ScrapeConfig to reference that file and scrape the associated targets:

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: ScrapeConfig
metadata:
  name: scrape-config-example
  namespace: my-namespace
  labels:
    app.kubernetes.io/name: scrape-config-example
spec:
  fileSDConfigs:
    - files:
        - /etc/prometheus/configmaps/scrape-file-sd-targets/targets.yaml

>       refreshInterval: 15s

Author: done 👍

```

## `http_sd`

`http_sd` works the same way as the `file_sd` but requires an endpoint providing that data rather than a file. For instance:

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: ScrapeConfig
metadata:
  name: scrape-config-example
  namespace: my-namespace
  labels:
    app.kubernetes.io/name: scrape-config-example
spec:
  httpSDConfigs:
    - url: http://my-external-api/discovery
      refreshInterval: 15s
```

## `kubernetes_sd`

`kubernetes_sd` allows retrieving scrape targets from Kubernetes' REST API. Here is an example that selects nodes in the kubernetes cluster:

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: ScrapeConfig
metadata:
  name: scrape-config-example
  namespace: my-namespace
  labels:
    app.kubernetes.io/name: scrape-config-example
spec:

>   kubernetesSDConfig:

Reviewer: We would need a TLSConfig and Authorization section SC spec 
here is the example I tested https://github.com/prometheus-operator/prometheus-operator/pull/5675#issuecomment-1608867483

Also cluster role binding was required for prometheus so that prometheus serviceAccount has access to list Node objects otherwise prometheus errors about access issue so we should mention that as well

Author: Following I will remove `kubernetes_sd` to keep it short and we can figure out a specific documentation for it as there is more context than that (e.g. "How to scrape an exporter located on all the kubernetes worker").
